game-core/src/main/java/games/strategy/triplea/ui/MovePanel.java

```diff
@@ -607,6 +615,48 @@ private void selectEndPoint(final Territory territory) {
           updateRouteAndMouseShadowUnits(null);
           release();
         }
+
+        private boolean confirmEndPoint(final Territory territory) {
+          final PlayerId player = getCurrentPlayer();
+          final boolean willStartBattle =
+              Matches.territoryHasUnitsOwnedBy(player)
+                  .negate()
+                  .and(Matches.territoryHasEnemyUnits(player, getData()))
+                  .test(territory);
+          if (!willStartBattle) {
+            return true;
```

Comment:
nit, the multiple return points can make things a bit harder to follow. We're also hiding perhaps too many details in `confirmEndpoint` and arguably a mixture of abstraction layers.

["Imperative shell - functional core"](https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell)  can perhaps help here. The idea is you want an outer layer that has the control flow logic which invokes a functional (core) business layer.

In this example it would look something like this (at L548):
```
if(willStartBattle(territory) && hasPossibleScramblers(territory) && !userConfirms(territory)) {
  return
}
```

To explain a bit further, the above keeps the control flow to one layer, makes it readily more obvious what is happening at that level, and then each part of the condition becomes functional and single purposed. That also helps solve that `confirmEndPoint` does a bit more than confirm, it checks to see if a confirmation is needed and then shows the confirmation dialog.

Last comment, for a lot of bonus points, we are wanting to extract logic from Swing UI code, particularly in light of our desired migration to JavaFX, but in general we want to fix the mixing of logic with UI code.

As an example, that last variant would look like this:
```
if(ScramblerLogic.territoryEligibleForScrambling(territory) && !userConfirms(territory)) {
  return
}
```

Reply:
The ScramblerLogic.territoryEligibleForScrambling() suggestion does not work, since the warning dialog actually shows the units to scramble, which need to be passed to userConfirms().

However, I did make a helper function for willStartBattle() and another for showScrambleWarningAndConfirmMove(). I agree that with this change, the control flow in confirmEndPoint() is much easier to follow - thanks for the suggestion.